### ISSUE TITLE
Create package naming guidance for community members

### ISSUE THREAD

**Issue by codefromthecrypt** [open-telemetry/community#2194]
It would be a good idea to have package naming guidance available for community members.

Here's an example guidance we could come up with or revise towards that end:

```markdown
### Repositories with global namespaces

Do not prefix a 3rd party package with `opentelemetry-` in a global namespace such as [pipi](https://pypi.org/). Instead, prefix it with the 3rd party. For example, if your package is owned by elastic call it `elastic-opentelemetry` not `opentelemetry-elastic`.

If you are packaging on behalf of another, prefix to make this clear. For example, if you are the vendor "metrix" and you are publishing an opentelemetry package for "openai", call it "metrix-openai" or "metrix-opentelemetry-openai" not "opentelemetry-openai" or "openai-opentelemetry".

These guidelines make it clear who the maintainers are, and also frees up the same package name for both OpenTelemetry and target maintainers. Regardless, when documenting a 3rd party package, do not call it "the official OpenTelemetry" or "the standard OpenTelemetry" as it can create confusion with packages produced by OpenTelemetry.
```

The aim is that by the time someone becomes a maintainer or SIG lead, they already know both know the rules and also don't violate them in their own 3rd party works. In other words, this document is both to provide coherent and succinct advice, and also prevent accidents from repeating.

One way to apply such a document is to make it a part of the process to promote inside our org. For example, to be an approver, you must have at least read this advice. Hopefully, by reading this, folks will voluntarily comply with it, and set a good example for others.

See also related issues:
* https://github.com/open-telemetry/community/issues/1096

**Comment by svrnm** [comment]
I wonder where something like this should live, is this a spec thing, or something we should have somewhere else? spec seems to be the easy thing to do?

On the question itself, I agree, that there should be some clarity around naming packages for 1st and 3rd party packages, this is also somehow related to [this discussion](https://github.com/open-telemetry/opentelemetry-specification/issues/4089), where we try to make it easier to distinguish between different "kinds" of instrumentation.

Enforcing/Encouraging this package naming guidelines is a different subject and I am not sure how far we want to go with that. What I would like to do (but we don't do today) is enforce it in the registry.

@codefromthecrypt would you be able to bring a list of existing packages from community and 3rd parties for 1 or 2 languages such that we can get an understanding of the problem.

**Comment by codefromthecrypt** [comment]
> @codefromthecrypt would you be able to bring a list of existing packages from community and 3rd parties for 1 or 2 languages such that we can get an understanding of the problem.

Someone beat me to it:

https://github.com/open-telemetry/opentelemetry.io/pull/4836

**Comment by svrnm** [comment]
> > @codefromthecrypt would you be able to bring a list of existing packages from community and 3rd parties for 1 or 2 languages such that we can get an understanding of the problem.
> 
> 
> 
> Someone beat me to it:
> 
> 
> 
> https://github.com/open-telemetry/opentelemetry.io/pull/4836

I don't think those 2 are related. I was wondering about packages in 3rd party registries like pypi packaging npm etc

**Comment by codefromthecrypt** [comment]
right good point. Those adding themselves to our registry would be a subset of those in 3rd party registries. That said, our registries are easier to search ;)

Specifically, I think pipi is the more easy to misuse of languages I've worked with recently. There are a myriad of packages misused, but parsing out a comprehensive list requires some python engineering. For example, https://pypi.org/search/?q=opentelemetry-instrumentation- returns over 3.3k results. Most who have attempted anything better abandoned their html scrapers.

Anyway, the first page varies of "opentelemetry-instrumentation-", but probably just by odds you'll find one using this convention which isn't otel managed.

For example, the first page hit for me includes many upstream packages, but also many not. Easiest way to see is by version skew
opentelemetry-instrumentation-replicate by @nirga etc at traceloop/openllmetry
opentelemetry-instrumentation-together      ^^
opentelemetry-instrumentation-alephalpha  ^^
opentelemetry-instrumentation-anthropic    ^^
opentelemetry-instrumentation-bedrock      ^^

second page is similar, but highlights that while openllmetry are a very large count, it isn't just openllmetry packages:
opentelemetry-instrumentation-chromadb by @nirga etc at traceloop/openllmetry
opentelemetry-instrumentation-cohere            ^^
opentelemetry-instrumentation-haystack         ^^
opentelemetry-instrumentation-langchain        ^^
opentelemetry-instrumentation-llamaindex      ^^
opentelemetry-instrumenation-digma          by @shaykeren at digma
opentelemetry-instrumentation-discordpy  by @Cookie at Idegrity
opentelemetry-instrumentation-faststream by @ashambalev
opentelemetry-instrumentation-kser            by @cdumay 
opentelemetry-instrumentation-kstreams    by @woile at kpn

Due to finding so many in only 2 pages of search, I think basically the word isn't out. Many repos don't have this issue because rules are enforced prior to publication. Pipi is basically easy to misuse. While folks like openllmetry are very active and may be able to easily rename to openllmetry-instrumentation- etc. it could be worse in other scenarios. Such as an abandoned library which is using a name that ends up in contrib. Then, it gets very messy.

So, basically I think we should create the guidance to help make it easier to understand the impact of using the reserved naming conventions, even to know it is reserved. Then, people like me who aren't quite members yet, when we go to become one, we can learn about this topic, and help steer community in a way that doesn't result in lockups.

On the bigger ask about like knowing all use, I think that's scripting investment and since it is really CNCF at the end of the day, perhaps there are some resources out there to help with this. Then even if you aren't a member of otel, or advised by one, people can find out the glitch quickly before renaming packages becomes annoying for their users.

Hope this helps!

**Comment by codefromthecrypt** [comment]
It was wrong for me to say "misuse" above wrt folks using the same conventions as the opentelemetry group. It was imprecise, but I meant no harm. I should have just "using the same conventions as opentelemetry maintained packages". I don't know a concise way to put that.

It is clear the group is interested in conventional enforcement and is attempting to clarify things. Until such time as "opentelemetry-instrumentation-" is documented (or someone can find one), we just have common concerns. This issue is in fact trying to clarify this.

For example,
* We do say that "otel." is [reserved](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/semantic-conventions.md#reserved-namespace) (as pointed out by @austinlparker)
* With [hex.pm](http://hex.pm/) we want to prevent people from using opentelemetry_ prefix unless in the opentelemetry (pointed out to me by @tsloughter)
* In namespaces like maven you can't easily publish anything that would cause significant confusion as the groupID would differ (my 2p)

All that said, I can't find anything that specifically says "opentelemetry-" literally in that combination of characters. The whole point of this issue is to make things clear and not implied. I used examples because I was asked about it. Please forgive my word mistakes as I would like to be able to help when asked without worrying minor word choice is problematic. Assume good intent!

**Comment by codefromthecrypt** [comment]
I originally opened this issue as I thought it should be knowable, so that at least approvers or maintainers should know that the ecosystem works best when folks don't clash with conventions used by our language SIGs.

I opened this issue because unlike ad-hoc oss, we have distinct governance and an effort to consolidate community practice. I thought we could leverage our infrastructure around giving good guidance to minimize tension on the community caused by package clashes. This was under the conjecture that having documented this, and put it in notes to read, clashes would be less likely to repeat or proliferate.

What I feel differently about now, is that this is a slippery slope to enforcement, or hard feelings. Even saying this stuff out loud feels like "picking on people" who make accidents or just feel like they prefer the names that we also use and shouldn't feel judged about that even if they do clash.

Maybe the cure is worse than the disease? These sort of practices aren't ours to allow or deny anyway.

In summary, I'm closing this out because basically I see no solution or action to come of this. I got carried away with documentation and lost that this is a touchy topic, and looking only at this from the POV of lockups in contrib misses most of it.

Cheers, over and out
